api/common/api-profile.go

```diff
@@ -0,0 +1,457 @@
+		// User has no permission to edit this field
+		// Component should not be aware of its existence
+		return nil
+	}
+	// By default, attributes are not shown for a given frontend
+	if frontend != "" && !attrb.IsAnnotationTrue(frontend) {
+		// Attribute is not declared to be used by the given frontend
+		return nil
+	}
+	var required = isAttributeRequired(attrb, requesterType)
+	return &ProfileAttributeRepresentation{
+		Name:        attrb.Name,
+		DisplayName: attrb.DisplayName,
+		Group:       attrb.Group,
+		Required:    &required,
+		Validations: ValidationsToApi(attrb.Validations),
+	}
+}
+
+func isAttributeRequired(attrb kc.ProfileAttrbRepresentation, requesterType string) bool {
+	if attrb.Required == nil {
+		return false
+	}
+	return validation.IsStringInSlice(attrb.Required.Roles, requesterType)
+}
+
+// ValidationsToApi converts KC validators
+func ValidationsToApi(validations kc.ProfileAttrbValidationRepresentation) map[string]ProfileAttrbValidatorRepresentation {
+	if validations == nil {
+		return nil
+	}
+	var res = make(map[string]ProfileAttrbValidatorRepresentation)
+	for k, v := range validations {
+		res[k] = ToValidator(v)
+	}
+
+	return res
+}
+
+// ToValidator converts a KC validator
+func ToValidator(validator kc.ProfileAttrValidatorRepresentation) ProfileAttrbValidatorRepresentation {
+	if validator == nil {
+		return nil
+	}
+	var res = make(ProfileAttrbValidatorRepresentation)
+	for k, v := range validator {
+		res[k] = v
+	}
+	return res
+}
+
+// GroupsToApi converts KC users profile groups to its API version
+func GroupsToApi(groups []kc.ProfileGroupRepresentation) []ProfileGroupRepresentation {
+	if groups == nil {
+		return nil
+	}
+	var res = []ProfileGroupRepresentation{}
+	for _, group := range groups {
+		res = append(res, ProfileGroupRepresentation{
+			Name:               group.Name,
+			DisplayHeader:      group.DisplayHeader,
+			DisplayDescription: group.DisplayDescription,
+			Annotations:        group.Annotations,
+		})
+	}
+	return res
+}
+
+// DefaultProfile returns a default profile
+func DefaultProfile(realmName string) (kc.UserProfileRepresentation, error) {
+	var res kc.UserProfileRepresentation
+	var err = json.Unmarshal([]byte(defaultUserProfile), &res)
+	return res, err
+}
+
+// ContainsFields interface
+type ContainsFields interface {
+	GetField(name string) interface{}
+	SetField(name string, value interface{})
+}
+
+// ValidateUser against a given profile
+func ValidateUser(profile kc.UserProfileRepresentation, input ContainsFields, frontendType string, requesterType string, checkMandatory bool) error {
+	for _, attrb := range profile.Attributes {
+		if !attrb.IsAnnotationTrue(frontendType) {
+			// Attribute is not supposed to be provided for this frontend type
+			input.SetField(*attrb.Name, nil)
+		} else {
+			var value = input.GetField(*attrb.Name)
+			if value == nil {
+				if checkMandatory && isAttributeRequired(attrb, requesterType) {
```

Comment:
In case you are in an update context, you may want to disable the "mandatory" checks as you know that provided nil values will keep the previous value (which has already been validated)